lib/history/sqlite/event.go

```diff
@@ -0,0 +1,98 @@
+package sqlite
+
+import (
+	"database/sql"
+	"time"
+
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+	"github.com/gravitational/satellite/lib/history"
+
+	"github.com/gravitational/trace"
+)
+
+// sqlEvent defines an sql event row.
+type sqlEvent struct {
+	ID        int            `db:"id"`
+	Timestamp time.Time      `db:"timestamp"`
+	EventType string         `db:"type"`
+	Node      sql.NullString `db:"node"`
+	Probe     sql.NullString `db:"probe"`
+	Old       sql.NullString `db:"oldState"`
+	New       sql.NullString `db:"newState"`
+}
+
+// newSQLEvent constructs a new sqlEvent from the provided TimelineEvent.
+func newSQLEvent(event *pb.TimelineEvent) (row sqlEvent, err error) {
+	row = sqlEvent{Timestamp: event.GetTimestamp().ToTime()}
+	switch t := event.GetData().(type) {
```

Comment:
This actually lends itself perfectly to interfaces - instead of exhaustively enumerating all known types (which would require changes with each new type added) factor this into an interface for events to implement and avoid the bogus `trace.BadParameter("unknown event type %T", t)`

Reply:
Still a little confused on how to structure the interfaces. Are you looking for something like this?

```golang
// SQLInserter inserts row into underlying SQL database.
type SQLInserter interface {
  // InsertRow prepares row to be inserted.
  InsertRow(SQLRow) error
  // ExecInsert inserts all prepared rows to underlying storage.
  ExecInsert() error
}

// SQLRow can be inserted into a SQL database.
type SQLRow interface {
  // Insert inserts row into storage.
  Insert(*sql.DB) error
}

// ProtoBuffer can be converted into a protobuf TimelineEvent.
type ProtoBuffer interface {
  // ProtoBuf returns event as a protobuf message.
  ProtoBuf() *pb.TimelineEvent, error
}
```
SQLInserter can be implemented with different sql databases.
We can then define separate structs for each event type and each event type implements `SQLRow` and `ProtoBuffer` to define their own `Insert` and `ProtoBuf` methods.